SI-7008 @throws annotations are now populated in reflect

xeno-by opened the pull request:
Runtime reflection in JavaMirrors previously forgot to fill in
@throws when importing Java reflection artifacts. Now this is fixed.

Note that generic exception types used in `throws` specifications will
be garbled (i.e. erased), because we don't use `getGenericExceptionTypes`
in favor of just `getExceptionTypes` to stay compatible with the behavior
of ClassfileParser. That's a bug, but a separate one and should be fixed
separately.

Also note that this commit updated javac-artifacts.jar, because we need
to test how reflection works with javac-produced classfiles. The sources
that were used to produce those classfiles can be found in the jar next
to the classfiles.

xeno-by committed:
SI-7008 @throws annotations are now populated in reflect

Runtime reflection in JavaMirrors previously forgot to fill in
@throws when importing Java reflection artifacts. Now this is fixed.

Note that generic exception types used in `throws` specifications will
be garbled (i.e. erased), because we don't use `getGenericExceptionTypes`
in favor of just `getExceptionTypes` to stay compatible with the behavior
of ClassfileParser. That's a bug, but a separate one and should be fixed
separately.

Also note that this commit updated javac-artifacts.jar, because we need
to test how reflection works with javac-produced classfiles. The sources
that were used to produce those classfiles can be found in the jar next
to the classfiles.

xeno-by commented:
review @gkossakowski @retronym @paulp

gkossakowski commented:
@xeno-by regarding `javac-artifacts.jar`, is there any reason you cannot follow the setup I used in #2021 where use `_1`, `_2`, etc. file suffixes to compile Java files first and use them on classpath for the test?

xeno-by commented:
@gkossakowski That's a good question. Back then I had problems with that, but I cannot remember their nature. Are `java` files in partests compiled by `javac` or by `scalac`?

gkossakowski commented:
@xeno-by: they are compiled with javac. Scalac can only parse Java files to populate symbol table for mixed compilation. It doesn't compile Java files :)

It would be useful to try. I prefer everything to be in source code. Eventually, we should get rid of `javac-artifacts.jar` altogether.

xeno-by commented:
I agree. If there are no problems, I don't see a reason why we should be roundabout with all those auxiliary jars.

xeno-by commented:
@gkossakowski Btw I noticed that your nearby pull request changes how `ClassfileParser` creates `@throws` annotations. This PR should do exactly the same. What do we do about this dependency?

gkossakowski commented:
@xeno-by: yes, I agree it should exactly the same. What problem do you see here? Or do you want to check if Scala runtime reflection produces the same result as Scala compile time reflection?

xeno-by commented:
I just wanted to know whether you're done with that PR, so that I can rebase off it and reuse the changed logic of `ClassfileParser` in `JavaMirrors`.

gkossakowski commented:
@xeno-by: Yes, I think it should be merged as-is. I'll ping @adriaanm to look at it ASAP.

xeno-by commented:
@gkossakowski Thanks!

xeno-by commented:
@gkossakowski @adriaanm There's a problem, though. Unfortunately, I cannot do `typer.packSymbols(cls.typeParams, cls.tpe)` in runtime reflection. What would you advise to do instead?

paulp commented:
I propose:

```

scala> import scala.reflect.runtime.universe._
import scala.reflect.runtime.universe._

scala> val clazz = typeOf[scala.throws[_]].typeSymbol.asClass
clazz: reflect.runtime.universe.ClassSymbol = class throws

scala> existentialAbstraction(clazz.typeParams, clazz.toType)
res0: reflect.runtime.universe.Type = throws[_ <: Throwable]
```

xeno-by commented:
@paulp done

gkossakowski commented:
@paulp: Interesting. Should I use the same thing in `ClassfileParser` in #2021 ?

xeno-by commented:
@gkossakowski I already did that - the logic is now shared between `JavaMirrors` and `ClassfileParser`.

gkossakowski commented:
@xeno-by: I didn't notice. Thanks!

BTW. I see that you still refer to java types from `javac-artifacts.jar`. The idea to compile everything from source didn't fly?

xeno-by commented:
Didn't have time for that yet. Will do that today or tomorrow.

xeno-by committed:
evicts javac-artifacts.jar

Apparently, the usual _1, _2, _3... naming scheme also works for java
files, which need to be compiled together with partests. This allows us
to get rid of javac-artifacts.jar.

xeno-by commented:
Done. That was easier than I thought :)

gkossakowski commented:
@xeno-by legitimate failure.

retronym commented on src/reflect/scala/reflect/internal/AnnotationInfos.scala:
```diff
@@ -33,6 +33,17 @@ trait AnnotationInfos extends api.Annotations { self: SymbolTable =>
       case ThrownException(exc) => exc
     }
 
+    def addThrowsAnnotation(excSym: Symbol): Self = {
```
Can we pick something more expressive for `excSym` / `excTpe`? e.g. `throwableSym/Tpe`?

xeno-by replied:
Done

retronym commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -610,11 +610,19 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
      */
     private def copyAnnotations(sym: Symbol, jann: AnnotatedElement) {
       sym setAnnotations (jann.getAnnotations map JavaAnnotationProxy).toList
+      // FIXME: we're not using getGenericExceptionTypes here to be consistent with ClassfileParser
```
When would this be fixed? Let's track it in a ticket.

xeno-by replied:
Done

retronym commented on src/reflect/scala/reflect/runtime/JavaMirrors.scala:
```diff
@@ -610,11 +610,19 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
+        case jconstr: jConstructor[_] => jconstr.getExceptionTypes.toList
+        case _ => Nil
+      }
+      jexTpes foreach (jexTpe => sym.addThrowsAnnotation(classSymbol(jexTpe)))
```
How are Scala classes containing `@throws` annotations treated? I can't figure out whether we pickle the annotation in addition to adding the exception to the signature. If we do, might we end up with duplicate annotations in runtime reflection? This warrants a test.

xeno-by replied:
Scala classes are unpickled with a different completer.

xeno-by replied:
I've added a test anyways :)

xeno-by committed:
pullrequest feedback

https://github.com/scala/scala/pull/2040

xeno-by committed:
reflecting @throws defined in Scala code

As per Jason's comment: How are Scala classes containing @throws annots
treated? I can't figure out whether we pickle the annotation in addition
to adding the exception to the signature. If we do, might we end up with
duplicate annotations in runtime reflection? This warrants a test.
See the context of the discussion here: https://github.com/scala/scala/pull/2040/files#r2874769.

No, we won't end up with duplicates, because classes defined in Scala
are loaded in a different completer. But I'll add a test - you can never
have too many of those.

xeno-by commented:
All done, Jason. Can I get my rubber stamp now? :)

retronym commented:
LGTM :ok: :+1:

xeno-by commented:
@adriaanm Why does the UI say "Good to merge", while there's pr-checkin-per-commit in flight?

adriaanm commented:
Because github only looks at the most recent status in a commit.
I'm working on reconfiguring jenkins so that we only have one job per commit now.
